### PR TITLE
Create cli tool, settings, and slurm utils

### PR DESCRIPTION
**Pull request by m-walters**
I haven't gone through and touched/refactored all the places that these `settings` would apply to (for example dry-run not writing anywhere). So we should patch this in as we go.

### CODE REVIEW

File: settings.py
Comment (mukobi) [inline settings.py:31]: I think we may just not need this class/file. How I tend to prefer to do this kind of thing is with a bunch of parameters defined with argparse, then load them with wandb.init()--see https://docs.wandb.ai/guides/track/config#set-the-configuration-with-argparse. Then, Weights & Biases will handle all the config saving and overriding for us.

**Review by mukobi — CHANGES_REQUESTED** [review]

File: README.md
Comment (m-walters) [reply README.md:8]: Yeah, the idea is that we don't need to be installing python notebooks and matplotlib etc when creating envs for program execution on cluster.

File: main.py
Comment (m-walters) [reply main.py:98]: Yeah that's what non-interactive was for, to allow a confirmation and clarification around creating/overwriting directories, but i'll scrap it

File: requirements.txt
Comment (m-walters) [reply requirements.txt:5]: Will merge when creating `toml` etc.

Comment (m-walters) [reply requirements.txt:5]: Alright

File: sbatch_run
Comment (m-walters) [reply sbatch_run:5]: I left this as small from my testing. `--mem` is memory per "node" which I imagine includes GPUs. Can leave out though, sure

File: settings.py
Comment (m-walters) [reply settings.py:31]: Wouldn't this require passing the config object to basically every function if they want to query any settings/configs? I think Rohan has already set up config parsing with OmegaConf

File: cli.py
Comment (m-walters) [reply cli.py:33]: Removed non-interactive

File: README.md
Comment (m-walters) [reply README.md:21]: hmm, I'd rather not hold up the PR configuring this as a package, so I'll remove this and we'll avoid nested imports until then

File: settings.py
Comment (mukobi) [reply settings.py]: `wandb.config` is a dict available globally, so we don't even have to pass it in. See https://docs.wandb.ai/guides/track/config

File: README.md
Comment (mukobi) [reply README.md:17]: Or since these are just for initial setup, you could have two separate setup scripts for conda and venv. Though that feels needlessly redundant.

File: sbatch_run
Comment (mukobi) [reply sbatch_run]: Node kinda means an individual computer. A node can multiple GPUs and CPUs attached to it. --mem still specifies the CPU RAM, in MB if you omit a unit. https://slurm.schedmd.com/srun.html#:~:text=to%20job%20allocations.-,%2D%2Dmem%3D%3Csize%3E%5Bunits%5D,-Specify%20the%20real Should just remove for now.

Comment (m-walters) [reply sbatch_run]: Sure, np!

File: README.md
Comment (mukobi) [inline README.md:14]: You probably should tell developers to install only in editable mode, so you can remove the first one here.

File: cli.py
Comment (mukobi) [inline cli.py:34]: Since we'll be trying many things, I'm not sure it makes sense to have many settings stored in config files. Would recommend not using OmegaConfig, instead defining all our arguments here in argparse. Can make them hierarchical, like `--model.path` and `--model.lora.rank`. Then, when loaded into `wandb.init(config=args)`, `wandb.config` will be a hierarchical object of all the settings. That also makes wandb sweeps over these settings nicer to work with.

File: settings.py
Comment (m-walters) [reply settings.py]: ok added

File: cli.py
Comment (m-walters) [reply cli.py:46]: But we define all out model parameters in these config files. Putting these all in through argv sounds super painful

File: README.md
Comment (m-walters) [reply README.md]: How about we leave this for a later PR and get this one merging in.

File: cli.py
Comment (mukobi) [reply cli.py:46]: If using config files, how do you specify changes for individual experiments (surely, we don't want a config file per experimental condition)? We'd set default values in cli.py for each of the values, so you'd only specify the things you want to be different from default. This makes it especially nice to run sweeps over config values https://docs.wandb.ai/guides/sweeps/walkthrough

Comment (m-walters) [reply cli.py:46]: Hmm..I'm not aware of a way to, through argparse, define an argument such as `"--model"` that adapts to users passing in new tokens like `--model.path=...`. Instead of spelling out every potential param, I have a `json` solution where `"--model"` is a string arg, but users can make the string a json object e.g. `"{'path': '/some/path'}"` and this will be loaded as a dict to conf.

It's worth noting that you can specify wandb sweeps as config files: https://docs.wandb.ai/guides/sweeps/define-sweep-configuration
so I think we should keep config loading as well.

Comment (mukobi) [reply cli.py:55]: Argparse doesn't adapt to new sequences of parameters, but we don't need that functionality. We'd just specify each of the configurable parameters to our experiments exactly once with default values and tooltips in cli.py. Using periods in the parameter values would be purely aesthetic to help separate different clusters of parameters from each other, though we might use some other syntax like underscores.

I'm down to try omegaconf, but it feels messy to have both that kind of a config file and other argparsed args. If you want to use omegaconf files, it might be better to use [hydra](https://hydra.cc/docs/intro/) to automatically merge in the default values from a config file and any updated values from the command line or a sweep. Hydra is kinda officially supported by wandb if you explicitly convert the config to a dict before wandb project initialization (https://docs.wandb.ai/guides/integrations/hydra). Otherwise, you can use OmegaConf.load(), OmegaConf.from_cli(), and OmegaConf.merge() together to implement hydra-like behavior of having default config file values and updated cli values.

File: configs/wandb_sample.yaml
Comment (mukobi) [inline configs/wandb_sample.yaml:1]: We probably shouldn't commit this example file to our repo. Unless this was accidentally included?

Also, I think it could be good to have our config file(s) and sweep files separate, e.g. in folders called `configs/` and `sweeps/`

File: cli.py
Comment (mukobi) [inline cli.py:94]: This config overriding will get annoying. Would suggest Hydra, omegaconf merging, or argparse default values. See [this other comment](https://github.com/mukobi/Unadaptable-Foundation-Models/pull/16/commits/36d103577ff5072b0f25294c545d2dff2e0b56de#r1558812665)

**Review by mukobi — CHANGES_REQUESTED** [review]
Just the omegaconf config parsing needs sorting out, then lgtm

File: README.md
Comment (mukobi) [inline README.md:48]: prob shouldn't say ArgParse here

File: README.md
Comment (mukobi) [inline README.md:138]: Are we ever gonna use one-off params? Possibly better to just require everything be defined in the config already, then we don't have to check if things are in the config first. Maybe the exception to this could be boolean flags we want to easily enable, like `+disablewandb`, but you could also just have that explicit in the configs as `wandb: enabled` and override it with `wandb=disabled`

File: config.py
Comment (mukobi) [inline config.py:13]: Eventually, we should probably use a logger passed in from main so we can filter and such. Should be fine for now.

File: main.py
Comment (mukobi) [inline main.py:94]: We prob won't need distributed training. Would suggest leaving this out until we prove that we need it.

File: README.md
Comment (m-walters) [reply README.md:138]: I was just describing built-in hydra functionality. We can't stop users from doing it, but i'll remove

**Pull request merged**
